File: frontend/lib/api/compliance.ts
import apiClient from "./client";

// ========== Stage 4: Privacy & CASL Wording ==========

export interface PrivacyWordingResponse {
  wording_type: string;
  version: string;
  content: string;
}

export const getPrivacyWording = async (wordingType: "privacy_policy" | "casl"): Promise<PrivacyWordingResponse> => {
  const response = await apiClient.get(`/compliance/privacy-wording?wording_type=${wordingType}`);
  return response.data;
};

export const confirmPrivacyWording = async (wordingType: "privacy_policy" | "casl"): Promise<void> => {
  await apiClient.post("/compliance/privacy-wording/confirm", { wording_type: wordingType });
};

// ========== Stage 4: Financial Setup ==========

export interface FinancialSetup {
  id: number;
  tenant_id: number;
  payroll_type: string | null;
  pay_schedule: string | null;
  wsib_class: string | null;
  is_confirmed: boolean;
  confirmed_at: string | null;
  created_at: string;
  updated_at: string;
}

export interface FinancialSetupCreate {
  payroll_type?: "weekly" | "bi_weekly" | "monthly" | "semi_monthly" | null;
  pay_schedule?: "monday" | "tuesday" | "wednesday" | "thursday" | "friday" | "last_day_of_month" | "first_day_of_month" | "fifteenth" | "last_friday" | null;
  wsib_class?: "retail" | "office" | "construction" | "manufacturing" | "healthcare" | "food_service" | "grooming" | "daycare" | "other" | null;
}

export const createFinancialSetup = async (data: FinancialSetupCreate): Promise<FinancialSetup> => {
  const response = await apiClient.post("/compliance/financial-setup", data);
  return response.data;
};

export const getFinancialSetup = async (): Promise<FinancialSetup> => {
  const response = await apiClient.get("/compliance/financial-setup");
  return response.data;
};

export const confirmFinancialSetup = async (): Promise<FinancialSetup> => {
  const response = await apiClient.post("/compliance/financial-setup/confirm", { confirm: true });
  return response.data;
};

// ========== Stage 4: HR Policies ==========

export interface HRPolicy {
  id: number;
  policy_type: string;
  title: string;
  content: string;
  is_required: boolean;
  version: string;
}

export const getHRPolicies = async (): Promise<HRPolicy[]> => {
  const response = await apiClient.get("/compliance/hr-policies");
  return response.data;
};

export const acknowledgeHRPolicies = async (policyIds: number[]): Promise<void> => {
  await apiClient.post("/compliance/hr-policies/acknowledge", { policy_ids: policyIds });
};

export const getAcknowledgementStatus = async (): Promise<{ has_acknowledged_all: boolean; user_id: number }> => {
  const response = await apiClient.get("/compliance/hr-policies/acknowledgement-status");
  return response.data;
};

// ========== Public endpoint for invitation ==========

export const getHRPoliciesForInvitation = async (token: string): Promise<{ policies: HRPolicy[] }> => {
  const response = await apiClient.get(`/onboarding/invitations/${token}/hr-policies`);
  return response.data;
};

File: backend/app/models/compliance.py
"""
Compliance Models - Stage 4

Privacy/CASL wording, Financial setup, HR policies, and employee acknowledgements.
"""
from datetime import datetime
from typing import Optional, List
from enum import StrEnum

from beanie import Document
from pydantic import Field


class PayrollType(StrEnum):
    WEEKLY = "weekly"
    BI_WEEKLY = "bi_weekly"
    MONTHLY = "monthly"
    SEMI_MONTHLY = "semi_monthly"


class PaySchedule(StrEnum):
    """Payday options."""
    MONDAY = "monday"
    TUESDAY = "tuesday"
    WEDNESDAY = "wednesday"
    THURSDAY = "thursday"
    FRIDAY = "friday"
    LAST_DAY_OF_MONTH = "last_day_of_month"
    FIRST_DAY_OF_MONTH = "first_day_of_month"
    FIFTEENTH = "fifteenth"
    LAST_FRIDAY = "last_friday"


class WSIBClass(StrEnum):
    """Predefined WSIB classes."""
    RETAIL = "retail"
    OFFICE = "office"
    CONSTRUCTION = "construction"
    MANUFACTURING = "manufacturing"
    HEALTHCARE = "healthcare"
    FOOD_SERVICE = "food_service"
    GROOMING = "grooming"
    DAYCARE = "daycare"
    OTHER = "other"


class FinancialSetup(Document):
    """Financial setup configuration for tenant."""
    
    tenant_id: str = Field(..., index=True, unique=True)
    payroll_type: Optional[PayrollType] = None
    pay_schedule: Optional[PaySchedule] = None
    wsib_class: Optional[WSIBClass] = None
    is_confirmed: bool = Field(default=False)  # Must confirm before activation
    confirmed_at: Optional[datetime] = None
    confirmed_by_user_id: Optional[str] = Field(default=None, index=True)
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)

    class Settings:
        name = "financial_setups"
        indexes = [
            "tenant_id",
            "confirmed_by_user_id",
        ]


class HRPolicyType(StrEnum):
    """Predefined HR policy types."""
    HEALTH_SAFETY = "health_safety"
    HARASSMENT = "harassment"
    TRAINING = "training"


class HRPolicy(Document):
    """HR policy templates (shared across all tenants)."""
    
    policy_type: HRPolicyType = Field(..., index=True, unique=True)
    title: str
    content: str  # Policy text content
    is_required: bool = Field(default=True)  # Required for all employees
    version: str = Field(default="1.0")
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)

    class Settings:
        name = "hr_policies"
        indexes = [
            "policy_type",
        ]


class PolicyAcknowledgement(Document):
    """Employee acknowledgement of HR policies."""
    
    user_id: str = Field(..., index=True)
    policy_id: str = Field(..., index=True)
    acknowledged_at: datetime = Field(default_factory=datetime.utcnow)
    ip_address: Optional[str] = None
    user_agent: Optional[str] = None

    class Settings:
        name = "policy_acknowledgements"
        indexes = [
            "user_id",
            "policy_id",
            ("user_id", "policy_id"),  # Compound unique index
        ]


class PrivacyWording(Document):
    """Hardcoded privacy policy wording (versioned)."""
    
    tenant_id: str = Field(..., index=True)
    wording_type: str = Field(..., index=True)  # "privacy_policy" or "casl"
    version: str = Field(default="1.0")
    content: str  # Hardcoded content
    confirmed_at: Optional[datetime] = None
    confirmed_by_user_id: Optional[str] = Field(default=None, index=True)
    created_at: datetime = Field(default_factory=datetime.utcnow)

    class Settings:
        name = "privacy_wordings"
        indexes = [
            "tenant_id",
            "wording_type",
            "confirmed_by_user_id",
        ]
